src/puppetlabs/puppetdb/utils.clj

```diff
@@ -264,6 +264,49 @@
       (apply assoc m new-kvs)
       m)))
 
+(defn maybe-strip-escaped-quotes
+  [s]
+  (if (= \" (first s) (last s))
+    (subs s 1 (dec (count s)))
+    s))
+
+(defn parse-matchfields
+  [s]
+  (string/replace s #"match\((\".*\")\)" "$1"))
+
+(defn parse-indexing
+  [s]
+  (string/replace s #"\[(\d+)\]" ".$1"))
+
+(defn split-indexing
+  [path]
+  (flatten
+    (for [s path]
+      (if (re-find #"\[\d+\]$" s)
+        (-> s
+            (string/split #"(?=\[\d+\]$)")
+            (update 1 #(Integer/parseInt (subs % 1 (dec (count %))))))
+        s))))
+
+(defn dotted-query->fact-path
+  "Splits a string on . unless the . is contained within a quoted substring."
+  [string]
+  (loop [[s & splits] (string/split string #"\.")
+         temp []
+         result []
+         quoted false]
+    (if (nil? s)
+      result
+      (if quoted
```

Comment:
What do you think about splitting this into two functions, one that does some loop recur here, and another which handles the quoted case? something like

```
(defn dotted-query->fact-path
  "Splits a string on . unless the . is contained within a quoted substring."
  [string]
  (loop [[s & splits] (string/split string #"\.")
         result []]
    (if (nil? s)
      result
      (if (and (= \" (first s)) (or (= (count s) 1) (not (= \" (last s)))))
          (let [[result rest] (handle-quoted-bs [s] splits)]
             (recur rest result))
          (recur splits (conj result s)))))))
```